src/nanoarrow/array_inline.h

```diff
@@ -0,0 +1,246 @@
+  int result;
+  int64_t zero = 0;
+
+  for (int i = 0; i < 3; i++) {
+    if (private_data->layout.buffer_type[i] == NANOARROW_BUFFER_TYPE_OFFSET &&
+        private_data->layout.element_size_bits[i] == 64) {
+      result = ArrowBufferAppend(ArrowArrayBuffer(array, i), &zero, sizeof(int64_t));
+      if (result != NANOARROW_OK) {
+        return result;
+      }
+    } else if (private_data->layout.buffer_type[i] == NANOARROW_BUFFER_TYPE_OFFSET &&
+               private_data->layout.element_size_bits[i] == 64) {
+      result = ArrowBufferAppend(ArrowArrayBuffer(array, i), &zero, sizeof(int32_t));
+      if (result != NANOARROW_OK) {
+        return result;
+      }
+    }
+  }
+
+  return NANOARROW_OK;
+}
+
+static inline ArrowErrorCode ArrowArrayFinishBuilding(struct ArrowArray* array,
+                                                      char shrink_to_fit) {
+  struct ArrowArrayPrivateData* private_data =
+      (struct ArrowArrayPrivateData*)array->private_data;
+
+  if (shrink_to_fit) {
+    // TODO: implement
+    return EINVAL;
+  }
+
+  // Make sure the value we get with array->buffers[i] is set to the actual
+  // pointer (which may have changed from the original due to reallocation)
+  for (int64_t i = 0; i < 3; i++) {
+    private_data->buffer_data[i] = ArrowArrayBuffer(array, i)->data;
+  }
+
+  return NANOARROW_OK;
+}
+
+static inline ArrowErrorCode ArrowArrayAppendNull(struct ArrowArray* array, int64_t n) {
+  struct ArrowArrayPrivateData* private_data =
+      (struct ArrowArrayPrivateData*)array->private_data;
+
+  if (n == 0) {
+    return NANOARROW_OK;
+  }
+
+  if (private_data->storage_type == NANOARROW_TYPE_NA) {
+    array->null_count += n;
+    array->length += n;
+    return NANOARROW_OK;
+  }
+
+  int result;
+
+  // Append n 0 bits to the validity bitmap. If we haven't allocated a bitmap yet, do it
+  // now
+  if (private_data->bitmap.buffer.data == NULL) {
+    result = ArrowBitmapReserve(&private_data->bitmap, array->length + n);
+    if (result != NANOARROW_OK) {
+      return result;
+    }
+
+    ArrowBitmapAppendUnsafe(&private_data->bitmap, 1, array->length);
+    ArrowBitmapAppendUnsafe(&private_data->bitmap, 0, n);
+  } else {
+    result = ArrowBitmapReserve(&private_data->bitmap, n);
+    if (result != NANOARROW_OK) {
+      return result;
+    }
+
+    ArrowBitmapAppendUnsafe(&private_data->bitmap, 0, n);
+  }
+
+  // Add appropriate buffer fill
+  struct ArrowBuffer* buffer;
+  int64_t size_bytes;
+
+  for (int i = 0; i < 3; i++) {
+    buffer = ArrowArrayBuffer(array, i);
+    size_bytes = private_data->layout.element_size_bits[i] / 8;
+
+    switch (private_data->layout.buffer_type[i]) {
+      case NANOARROW_BUFFER_TYPE_NONE:
+      case NANOARROW_BUFFER_TYPE_VALIDITY:
+        continue;
+      case NANOARROW_BUFFER_TYPE_OFFSET:
+        // Append the current value at the end of the offset buffer
+        for (int64_t j = 0; j < n; n++) {
+          result = ArrowBufferAppend(
```

Comment:
nit: might be good if this could be a single reserve + memset instead of looping